### ISSUE TITLE
Veracrypt is not accessible with Orca on Debian Jessie

### ISSUE THREAD

**Issue by Nardol** [veracrypt/VeraCrypt#171]
Hello,

When I start VeraCrypt GUI, Orca only says "Veracrypt inaccessible".
The only solution to use Veracrypt for the moment is to use in with console only version, which is not "good" for a beginer for example.
Could it be possible to solve this issue? Unfortunately I have no solution to propose.
We have not this issue on Windows, VeraCrypt is fully accessible since application key on drive list has been implemented.

**Comment by alt3r-3go** [comment]
As it's been a while, @pzajda (or anyone else reading this and interested) - is this still actual?

**Comment by pzajda** [comment]
Yes it is still actual, and in fact even GUI installer is not accessible now.
I have a Window titled "xmessage" where Orca cannot read the content at all.

**Comment by alt3r-3go** [comment]
Got it, thanks, I'll check this out then. Just to make sure I'm trying it in the same environment you have - what Linux distro/version and which Orca version (number, self-installed or from the distro repos, etc) do you use?

@idrassi, feel free to assign this one to me.

**Comment by pzajda** [comment]
Thanks a lot.
I use Debian Stretch 64'bits, with Orca 3.32.0.
It is not the Debian version, so I also tested with Orca Master with same result.

I hop I don't forget any essential information.

**Comment by alt3r-3go** [comment]
Thanks, yep, that should be good enough for starters.

**Comment by alt3r-3go** [comment]
All right, I have a test setup in place - sorry it took me a while. Debian Stretch with all current updates, Orca from the Debian repo (version 3.22.2), VeraCrypt 1.23-hotfix2 (latest released).

What I observe:
1. The "GUI" installer is indeed not working with Orca - but I'm not sure how it ever worked, AFAICS it's essentially the console installer that opens up xterm/xmessage windows with text in it. Orca apparently doesn't work with that. Let's leave this aside for now, as this would be a separate (and probably bigger) thing to fix.
1. The app GUI itself - it does work for me, in the main window Orca reads everything (menu bar, all the buttons) except the main "table" with the drive numbers and mounted volumes (just says "panel" there). The wizard windows seem to be a bit worse - it reads buttons and fields if I switch them using Tab key (and announces the one selected by default), but it doesn't read the text labels with descriptions, unless I use Orca's "flat review".

So focusing on the app itself for now - as I'm not sure what's expected and what's not in terms of Orca's ability to read it, would you please elaborate whether you think it should be able to read the static text labels by itself (in wizards)? The table (which is calls "panel") should probably be read, judging by the Orca's configuration window, where it reads it okay, but I couldn't quickly find a similar application that uses those static labels to convey the information to the user.

**Comment by alt3r-3go** [comment]
> Unlike you, I cannot use the menu bar nor with alt or F10
Oh, I see now, I used mouse and that's of course not the way one'd use it with Orca. Indeed when pressing Alt it only activates the keyboard accelerator letters in the menu, but doesn't activate menu itself. I'll see how it could be changed to activate the menu, so that you could then move using keyboard arrows. Just as a reference, if it helps you or anyone else in the meanwhile: Alt+V is Volumes menu, Alt+F is Favorites, Alt+O is Tools, Alt+G is Settings and Alt+H is Help.

> The "panel" if I understand you is the list of mounted volume, could be good if it was accessible
Yep, that's what I meant. I'll look into making it accessible as well. The benchmark results window actually uses the same control, so the same solution should work for both.

> I had a strange behavior with the Display PIM checkbox, it has began non-focusable
Do you mean "Use PIM" one? I don't see a "Display" one. If yes, do you mean the volume creation wizard or somewhere else? For me the "Use PIM" checkbox is announced as "Empty or zero for default iterations" instead of "Use PIM". The text it announces is the label that's shown if you actually click it, so this is broken as well anyway, but if you mean something else - please let me know.

All in all, looks like there's quite some work to do, but I do want to fix that. Will need to read up on GNOME accessibility guidelines to make sure I understand what the "right way" is, but that's not a problem, just time needed.

As for the installer - I think the proper way to fix this would be to simply ditch that custom installer at all (I don't see any value it brings) and package VeraCrypt using normal mechanisms for a couple of major distros + maybe Flatpak for the rest, so that it could be picked up by e.g. Debian/Ubuntu Software GUI tool and so on.

**Comment by pzajda** [comment]
> Do you mean "Use PIM" one? I don't see a "Display" one. If yes, do you mean the volume creation wizard or somewhere else? For me the "Use PIM" checkbox is announced as "Empty or zero for default iterations" instead of "Use PIM". The text it announces is the label that's shown if you actually click it, so this is broken as well anyway, but if you mean something else - please let me know.

We talk about the same thing.
Maybe I had a confusion but when I checked this checkbox, after that I had a checkbox to display the PIM.

Thanks a lot for you interest in accessibility of VeraCrypt GUI.

**Comment by alt3r-3go** [comment]
Just wanted to put a small update in (of course Murphy's law wanted that my free time evaporated right after I started looking into this, so I had to put it aside for a while). I'm currently looking into using Accercizer for evaluating the VC's accessibility to create a plan of action for fixing. And as I've never used or seen that tool before and it seems to be quite unfinished, it's an interesting challenge to say the least :) but I'm progressing.

**Comment by alt3r-3go** [comment]
As this seems to be a bigger task and I'm really short on free time recently, I've started splitting it into bite-size chunks so that it's manageable, as I still wanna help. Right now I'm focusing on making the main menu bar accessible and I'd be grateful for some clarification from @idrassi on the below.

As I already mentioned, pressing Alt or F10 on Windows _does_ activate the menu bar and the first menu item there (Hotkeys in case of Windows). Almost the same (only with F10, not Alt) - as I now see - happens with the version packaged into DEB. The version packaged into traditional installer doesn't respond to either Alt or F10 (and that's what @pzajda originally reported).

Now, the DEB version is apparently linked against wxWidgets lib while the installer one isn't (only native GTK+ libs AFAICS there). So my question would be - how exactly that traditional installer version is built (guess - doe it use `WXSTATIC=1` as outlined in the readme?) and do you think it makes any sense to dig into that for fixing or we should just ditch that installer and provide DEB + RPM packages (and link against wxWidgets lib) as discussed elsewhere?

**Comment by alt3r-3go** [comment]
Okay, never mind the first question, I se the "traditional" installer is indeed built with static mode set. The second is still actual, however I now realize we will need to understand what's going on in the static version anyway, for those who need it, regardless of whether we switch to dynamic linking as the main build type. Will look into that.

**Comment by alt3r-3go** [comment]
So I've built VC using wxWidgets 3.0.4 sources (clone from their github repo, 3.0.4 tag), using `WITHGTK3` option and the menu is being activated when using F10 key. So there must be some difference between what I did and what the traditional installer contains. @idrassi, could you please comment on this one?

**Comment by idrassi** [comment]
@alt3r-3go thank you this analysis and I apologize for not giving feedback earlier.
The traditional is linked statically wit wxWidgets 3.0.4 but using GTK2 and not GTK3 because the traditional installer is built under CentOS 6 which only has GTK2. CentOS 6 is used in order to have a binary that works across all Linux distributions. 

Thus, the issue may be coming from the use of GTK2 in a Desktop environment that uses GTK3.

Currently, the only VeraCrypt that are linked against GTK3 (through `libwxgtk3.0-gtk3-0v5` package) are those for Debian 10 and for Ubuntu versions from 18.04 upwards.

Does linking against GTK3 (either using static wxWidgets or using the system  `libwxgtk3.0-gtk3-0v5`) fixes the accessibility issue?

**Comment by alt3r-3go** [comment]
No worries and thanks for clarifying! Using GTK3 doesn't fix all the problems, but certainly makes things better - main menu is now accessible, at least, and I think in other places it's a bit more consistent as well (haven't checked in detail yet). I'll be looking into the next pieces, but I think the menu-related one is clear - moving to GTK3 is a solution here.

**Comment by alt3r-3go** [comment]
@Nardol I've been looking further into this and experimenting - and it looks like the best advice right now is to actually upgrade to Debian 10 or any newer distro with newer GNOME. It makes a difference - even though that main menu activation is just a matter of not using a generic installer and seems to work even on Debian 9, on a newer one (e.g. Fedora 30/31 I tested on) Orca's so called "flat mode" seems to work much better and reads everything properly, much like on Windows. On Debian 9 though, even compiled with GTK3, it doesn't work properly - announces only some words and not full phases, or none at all.

So to me it looks like they've been updating the accessibility infra in GNOME and it did visibly become better. What do you think?

**Comment by Nardol** [comment]
An upgrade of my system is planned on January. Anyway, after asking someone it looks like your observations are confirmed, accessibility components have ben largely improved.
The installer seems to be the only remaining problem if I understand you right.

**Comment by alt3r-3go** [comment]
Okay, that sounds good then - let's see if upgrade solves it fully for you and if not, please feel free to revive this one. One thing I think I still would want to look into is that "table" with mount and volume paths - I see in other programs this sort of controls is properly read by Orca (even on Debian 9), so there seems to be some room for improvement regardless of the OS/GNOME accessibility infra updates.

As for the installer - these days there are proper DEB packages for Debian 9 and 10 (and osme other distros), which are installable using standard `apt` tool. It's command-line, of course, but with installer it's also necessary to start it somehow, same as initiate the package installation. So I wouldn't use the installer at all, unless you have a Linux distro where none of the pre-built packages work.

**Comment by Nardol** [comment]
Oops, since I am now on Debian 10 but did not test the graphical 
interface yet.

But I will test very soon.

Le 30/12/2020 à 09:03, stale[bot] a écrit :

**Comment by Nardol** [comment]
I've just tested Veracrypt 1.24Update7 with Debian Buster and Orca master by installing the deb package.
What I observe:
- In the preferences, when check boxes are part of a group, I have to use the flat review to know the label. An example: "Preserve timestamp of file container";
- Volume list is now accessible as an editable list box so it is confirmed;
- Benchmark results are still not readable using Orca;
- In the volume creation wizard, the label for the volume file path is not read by Orca except with flat review;
- Still in the wizard: it could be good if the text to move the mouse to improve cryptography was announced by Orca. Because for the moment we can click format without having done any mouse movement.

To summarize: accessibility is really better now even if there are still some point to improve.
